### PR TITLE
Made certificates cache themselves as media files + S3 support

### LINKED ISSUE #121: Make volunteer badges cache themselves instead of redrawing on request
Currently every time you request a volunteer badge, it draws the template and then serves the drawn image to the user. This happens every time someone queries the badge, making it inefficient.
This is especially crucial now that we want to reduce CPU consumption!

Make the images cache themselves as static files which will be served and add hooks to field changes to redraw & resave the image if necessary.

### PR DESCRIPTION
**Pull request by stavlocker**
- Made certificates cache themselves as media files
     - **Django currently serves the media files even in production - this behavior needs to be stopped when #125 is done and we're in AWS and using S3**
 - navigating to /clients/ now redirects to /clients/home/
 - Small improvements to UI and fragmentation of UI components (get_certificate.html)
 - Changed client/get_certificate/ to client/find_certificate/

Fixes #121

### CODE REVIEW

File: client/certificate_manager.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            font=font,
+            align='right'
+        )
+        photo.save(certificate.image_path, format='png')
+    finally:
+        if photo is not None:
+            photo.close()
+
+
+g_last_certificate_purge = None
```
Comment (itay-sho) [inline client/certificate_manager.py]: there are some problems with globals in django. you are not guarnteed that they are kept between requests, especially in production.

Comment (itay-sho) [reply client/certificate_manager.py]: It probably should be managed by the certificate model ('cached_certificate_image' field or something like that)

File: client/certificate_manager.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    today = date.today()
+    if g_last_certificate_purge is None or today > g_last_certificate_purge:
+        g_last_certificate_purge = today
+        for certificate in certificate_model.objects.filter(expiration_date__lt=date.today()):
+            try:
+                os.remove(certificate_model.IMAGE_PATH_FORMAT.format(certificate.id))
+            except OSError:
+                continue
+            certificate.delete()
```
Comment (itay-sho) [inline client/certificate_manager.py]: there is no way to make crons on django by default (I think)
django-cron might be your friend (disclaimer - never used it. do some research)

File: client/models.py [hunk 3]
```diff
@@ -84,6 +89,37 @@ class Volunteer(Timestampable):
...
+                if getattr(self, key) != value:
+                    # A change occurred in one of the fields listed in the dict - update valid certificates images
+                    for certificate in self.get_active_certificates():
+                        client.certificate_manager.create_certificate_image(certificate)
+
+                    break
+
+        super().save(*args, **kwargs)
+        self._update_original_values()
+
```
Comment (itay-sho) [inline client/models.py]: is this related to the issue?

Comment (itay-sho) [reply client/models.py]: & I didn't understand what did you attempted to do here

File: client/models.py [hunk 4]
```diff
@@ -92,7 +128,11 @@ def get_or_generate_valid_certificate(self):
         return certificate
 
     def get_active_certificates(self):
-        return self.certificates.filter(expiration_date__gte=date.today())
+        """
+        Returns the active certificates a volunteer has, ordered by expiration date (first certificate has
+        most time to live)
+        """
+        return self.certificates.filter(expiration_date__gte=date.today()).order_by('-expiration_date')
```
Comment (itay-sho) [inline client/models.py]: cool

File: client/templates/components/get_certificate.html [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<h3>תג זה הינו הכרחי עבור ביצוע ההתנדבות - שמור את התג בטלפון שלך</h3>
+<img src="{{ certificate.image_url }}" alt="volunteer tag" width="400" />
+<br />
+<a href="{% url 'download_certificate' pk=certificate.id %}" download>הורדת תג</a>
```
Comment (itay-sho) [inline client/templates/components/get_certificate.html]: is it used more then once? otherwise merge it into thanks_volunteer

**Comment by itay-sho** [comment]
We are probably going to move to AWS faster than I've expected. let's make it S3-proof.
keep in mind to make non-s3 friendly as well, so that it'll work when working locally

I've seen this article that explains how to take static & media files from S3, can you look at that?
https://simpleisbetterthancomplex.com/tutorial/2017/08/01/how-to-setup-amazon-s3-in-a-django-project.html

File: client/templates/components/get_certificate.html (hunk 5 above)
Comment (stavlocker) [reply client/templates/components/get_certificate.html:4]: Yes, in thanks_volunteer and find_volunteer

File: client/models.py (hunk 3 above)
Comment (stavlocker) [reply client/models.py:122]: Yes: In order to check if one of the cruical fields (tz_number, first_name, last_name) changed, I'm storing the original value in the field, and before each save I check if the current value is the same as the original one. If it isn't, it means that the field change and the certificates should be reloaded. After this I re-update the original values so it will work again on the next save :)

**Comment by itay-sho** [comment]
this PR relays on #157

**Comment by stavlocker** [comment]
This was way more difficult than I expected!
I added S3 integration. Django serves static & media files in development, in production it does so from the S3 server.

* In deployment this requires 3 environment variables to specify the S3 credentials
* This requires migration (added an image field to VolunteerCertificate)

File: client/models.py [hunk 6]
```diff
@@ -119,8 +161,66 @@ def get_active_certificates(self):
...
+        s3 = boto3.client('s3')
+        return s3.generate_presigned_url(
+            'get_object',
+            Params={
+                'Bucket': 'levechad-media-bucket',
+                'Key': 'media/{}'.format(type(self).IMAGE_PATH.format(id=self.id)),
+                'ResponseContentDisposition': 'attachment;filename={}'.format(f'{self.id}.png'),
+            },
+            ExpiresIn=type(self).DOWNLOAD_LINK_EXPIRATION_SECONDS
+        )
```
Comment (itay-sho) [inline client/models.py:223]: why is it needed?

Comment (itay-sho) [reply client/models.py:223]: you can always use certificate.image.url

File: client/models.py [hunk 7]
```diff
@@ -84,6 +91,37 @@ class Volunteer(Timestampable):
...
+                if getattr(self, key) != value:
+                    # A change occurred in one of the fields listed in the dict - update valid certificates images
+                    for certificate in self.get_active_certificates():
+                        certificate.create_image()
+
+                    break
+
+        super().save(*args, **kwargs)
+        self._update_original_values()
+
```
Comment (itay-sho) [inline client/models.py:124]: this code might be looks to me too complex and likely to break in future changes...

File: client/templates/components/get_certificate.html [hunk 8]
```diff
@@ -0,0 +1,5 @@
+<h3>תג זה הינו הכרחי עבור ביצוע ההתנדבות - שמור את התג בטלפון שלך</h3>
+<img src="{{ certificate.image.url }}" alt="volunteer tag" width="400" />
+<br />
+<a href="{{ certificate.image_download_url }}">הורדת תג</a>
```
Comment (itay-sho) [inline client/templates/components/get_certificate.html:4]: why not?:
```suggestion
<a href="{{ certificate.image.url }}" download>הורדת תג</a>
```

File: client/urls.py [hunk 9]
```diff
@@ -20,5 +21,5 @@
     path('help/shopping', views.shopping_help),
     path('help/travel', views.travel_help),
     path('help/workers', views.workers_help),
-    path('volunteer_certificate_image/<int:pk>', views.volunteer_certificate_image_view, name='volunteer_certificate'),
+    # path('download_certificate/<int:pk>', views.download_certificate_view, name='download_certificate'),
```
Comment (itay-sho) [inline client/urls.py:24]: if this view is not required anymore, let's just remove it
```suggestion
```

File: levechad/settings.py [hunk 10]
```diff
@@ -151,10 +152,34 @@
...
 # Static files (CSS, JavaScript, Images)
 # https://docs.djangoproject.com/en/3.0/howto/static-files/
-
 STATIC_ROOT = os.path.join(BASE_DIR, 'staticfiles')
-STATIC_URL = "/static/"
-
-LOGIN_REDIRECT_URL = '/server'
+MEDIA_ROOT = os.path.join(BASE_DIR, 'mediafiles')
+
+if not DEBUG:
```
Comment (itay-sho) [inline levechad/settings.py:163]: let's be consistent with the rest of the page
```suggestion
if ENV != 'PRODUCTION':
```

File: client/models.py (hunk 6 above)
Comment (stavlocker) [reply client/models.py:229]: But that isn't a download link.
We used to use the download attribute on <a> tags, but this only works in Firefox and Chrome if the image is not from an external URL, which it now is (served from s3, not from django).
S3 has presigned URLs that expire after a while where you can give access to a resource using a URL, and you can also set the content disposition, which makes the users download the file.

File: client/models.py (hunk 7 above)
Comment (stavlocker) [reply client/models.py:124]: The only way I see this breaking is if someone refactors one of the fields, in which case they'd have to change this also (because it's a refactor). Could you explain how you see this break?

It's pretty generic - the only "hardcoding" you have here is the dict and the _update_original_values() function, but other than that it's generic.

This is the way to keep track of variables changing, because we have 3 it looks a bit more complex.

File: client/templates/components/get_certificate.html (hunk 8 above)
Comment (stavlocker) [reply client/templates/components/get_certificate.html:4]: I explained this above in my comment on image_download_url function

File: client/urls.py (hunk 9 above)
Comment (stavlocker) [reply client/urls.py]: I already did :)

**Comment by stavlocker** [comment]
@itay-sho All ready!
